### DISCUSSION TITLE
Adding Root Updates for Low Rank Updates and Adding Rows/Columns

### DISCUSSION THREAD

**Discussion by wjmaddox** [cornellius-gp/gpytorch#1497] [Ideas]
What would be the best way to incorporate functionality to update root decompositions of symmetric psd tensors such as:
`B = A + aa'` or adding new rows columns (e.g. https://github.com/cornellius-gp/gpytorch/blob/c74b8ed8d590fcb1d79808d9ee7cde168588ef99/gpytorch/models/exact_prediction_strategies.py#L183)?

I have a low rank implementation written as an `UpdatedRootLazyTensor` but it's suboptimal for general use because it's just a wrapper around a dense lazy tensor with an `update` function. Would it be best to just implement a `add_low_rank` to the root and a `add_new_rows` function analogously?

cc @Balandat @samuelstanton

**Comment by Balandat** [comment]
So from a user perspective a nice API would be if you had the ability to just add rows/columns to an existing LT and then automatically get the cheap update if you call `root_decomposition` on that. That might be a bit tricky to do though b/c it would require keeping a reference from the original LT to its root so it would introduce a "dependency" cycle.

**Reply by wjmaddox** [reply]
So #1499 addresses this by returning a stacked `CatLazyTensor` that has an overwritten root. Unfortunately, it's going to incur the cost upfront because doing it in lazy fashion is probably just going to eat a lot of "dependency" pointers.
